WDRT

cmichelenstrofer opened the pull request:
This pull request is for transferring the rest of the WDRT functionality to MHKiT. 

**Capabilities:**
- Buoy data I/O already in MHKiT
- Environmental contours are part of a separate PR
- [x] short-term extreme functions
- [x] sampling
- [x] extreme: full seastate
- [x] extreme: contour
- [x] extreme: MLER wave
- Equivalent fatigue load already in MHKiT

**Examples:**
- Contours example already in MHKiT.
- [x] short-term extreme example
- [x] Full sea state long-term extreme
- [x] Contour long term extreme 
- [x] MLER long term extreme
- Fatigue will be done later

**Tests:**
- [x] Tests

**Plan:**
- Carlos: short-term extreme, sampling, full seastate, contour approach
- Chris: MLER

cmichelenstrofer commented:
I've added the short-term extreme functionality and an example. I still need to add the function docstrings. 

Chris: can you review what I've done so far?

cmichelenstrofer commented:
This PR would close #22

hivanov-nrel commented on mhkit/loads/extreme.py:
```diff
@@ -0,0 +1,361 @@
+    ----------
+        x : np.array
+            Global peaks.
+
```
Need to add description of "threshold" input

hivanov-nrel commented on mhkit/loads/extreme.py:
```diff
@@ -0,0 +1,361 @@
+                out[x >= self.threshold] = 1. - (prop_pot * pot_ccdf)
+            return out
+
+    peaks = _Peaks(name="peaks", pot_distribution=pot, threshold=threshold)
```
Interesting use of a class inside a function. Is there a way to simplify this or is this the best way to do the computations?

cmichelenstrofer replied:
For some methods you simply fit a specific distribution (eg. Weibull). So you can  return a `scipy.stats distribution`. This is very convenient, you can do a lot with it (probability functions: PDF, CDF, PPF, etc; and metrics: expected value, etc). Other methods, however, are not as straightforward. The easiest solution I came up with was to subclass the `scipy.stats.rv_continuous` which can be  done by defining one of the PDF or CDF only, and you then get access to all the functionalities  of `scipy.stats distributions`. Else we would need to write our own implementation of these different functions and metrics. We should discuss this  more. 

- [Info on subclassing the `scipy.stats.rv_continuous`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.html)

cmichelenstrofer replied:
@ssolson we discussed this and think that subclassing `scipy.stats.rv_continuous` ([see documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.rv_continuous.html)) is acceptable here. What do  you think? Let me know if you want to discuss.

hivanov-nrel commented on mhkit/loads/extreme.py:
```diff
@@ -0,0 +1,361 @@
+    return peaks
+
+
+def peaks_to_ste(peaks_distribution, npeaks):
```
to keep consistency we should change the name to: ste_peaks

hivanov-nrel commented on mhkit/wave/resource.py:
```diff
@@ -158,19 +157,19 @@ def jonswap_spectrum(f, Tp, Hs, gamma=None):
 
 ### Metrics
-def surface_elevation(S, time_index, seed=123, frequency_bins=None,phases=None):
+def surface_elevation(S, time_index, seed=None, frequency_bins=None,phases=None):
```
The changes to this function are causing failures in the test. We should discuss what the correct algorithms are for computing the surface elevation.

cmichelenstrofer replied:
I will create a separate PR to address  this function. It has bugs in it and is very slow. My guess is the tests fail because of the seed not being hardcoded to 123. Or due to the truncation (bug). I will make sure to fix the tests as  well in that PR.

cmichelenstrofer commented:
@hivanov-nrel what is the plan/timeline for MLER?

cmichelenstrofer commented:
- [ ] Increase test coverage
- [ ] Pull in ssolson PR when it is ready

hivanov-nrel commented:
@cmichelenstrofer are we pretty close on this? What is left to do?

cmichelenstrofer commented:
STATUS OF WDRT MIGRATION TO MHKiT

@cmichelenstrofer  latest changes (@hivanov-nrel can you review?):
- Created a (non-hidden) 'waves.module`, rather than importing the functions into `waves.resource`
- Added 2 sampling functions for the full sea-state and contour approaches (`waves.contours`)
- Added a full seastate extreme function to ``loads.extreme`
- Created examples for the full seastate and contour approaches. Modified the MLER example name and intro so that these three examples are consistent (three different methods to do same thing)
- Merged the changes from the main repo including @ssolson's contours methods


@cmichelenstrofer  older changes (already reviewed by @hivanov-nrel )
- Create `loads.extreme` module and add short-term extreme methods
- Short-term extreme example


@hivanov-nrel  changes (see @cmichelenstrofer review to follow):
- add `loads.MLER` module with MLER functionality
- add MLER example
- add MLER tests


@ssolson (separate repo)
- Migrate all the different contour methods, create a single function which can call on any specified method
- Have the contour method output all relevant information (needed for sampling)
- Create contour example showing the different methods
- Add contour tests


TO DO
- [ ] Tests: short-term extreme methods
- [ ] Test: sampling: full seastate
- [ ] Test: sampling: contour approach
- [ ] Test: long-term extreme
- [ ] @hivanov-nrel review latest changes
- [ ] @hivanov-nrel finalize MLER based on review 
- [ ] @hivanov-nrel can you verify that the examples run (I do not have the ability to pull the buoy data using MHKiT, I had to manually download it). 
- [ ] Add a fatigue load example similar to the one in WDRT. MHKiT has an equivalent load function similar to the one in WDRT, so maybe there is no need to migrate WDRT's function. Someone should verify this. Even then, the WDRT example calculates a **long term fatigue load**, not just based on the observation time period. I.e. it converts from the 1 hour to a 1 year fatigue load. This capability should be migrated to MHKiT and showed in an example.
- [ ] Migrate missing functionalities from WDRT Contours  (can be done at a later date, I added place holder functions for these in `waves.contours`. This include calculating steepness curve and modifying contours appropriately, adding confidence intervals on the calculated contours using bootstrap, etc. These would need tests and examples too.

cmichelenstrofer commented:
@hivanov-nrel @ssolson @rpauly18 Unfortunately I won't be able to spend much more time on this. I provided a detailed summary (see previous comment) of the current status and what is left to do. I think we are most of the way there and should try to finish it soon (this coming release) rather than waiting for the next release. 

@ssolson the plan was for Adam to replace me in coding efforts starting this Q2, is he up to speed? Could he create the tests for the functions I wrote? 

@hivanov-nrel can you do the Fatigue task? 

@ssolson the missing contour capabilities (e.g. modifying based on steepness, adding confidence intervals) should be relatively easy and are related to the work you did in your branch. Can you or Adam take this one on?

cmichelenstrofer commented:
@hivanov-nrel My MLER Review

Some comments/suggestions: 

1.  My preference would be *not* to include the `readRAO` function for a few reasons
  - The response does not need to be a rigid body DOF motion for a single body device, it can be anything (relative motion between rigid bodies, PTO extension, tension on a mooring line, etc.)
  - It should be up to the user to convert their RAO (from whatever source) to the required input arrays, we should not be doing interpolation, or converting from period to frequency (what if the user already has frequency instead of period in their file?) 
 
2. The MLER function is not very large, maybe consider including it in the `loads.extreme` module where the functions for the other extreme loads approaches are. 

3. In the MLER example
  - I would keep going a little further and create the wave time-series, as this is the actual input (as boundary conditions) to the high fidelity (e.g. URANS) code
  - I would explain a bit more how this would then be used in a high fidelity code and what the design/extreme response is. 
  - you say *"we would like to find the wave profile that will generate a heave response of 1 meter for our WEC device"*. I am confused, how do you know the heave response beforehand? Isn't that the whole point of this example, to find the extreme heave response? I think you need to add more explanation on how the MLER method works,  what the goal is, and how to use the results.

cmichelenstrofer commented:
@hivanov-nrel The broken tests are fixed. Can you check if this latest push fixes the full sea state example for you? I will get started with the new tests.

hivanov-nrel commented:
Yup it looks like it all works! Okay I'm working on MLER and then short term extreme tests if you arent there already.

cmichelenstrofer commented:
I haven't gotten started on the new tests. Will update here as I complete them. I will mostly work on this in the afternoon, got a few morning meetings.

cmichelenstrofer commented:
@hivanov-nrel I think the MLER test is failing

cmichelenstrofer commented:
Error with line 176

```python
assert_frame_equal(mler_ts,self.mler_ts)
```

>======================================================================
FAIL: test_MLERexportTimeSeries (mhkit.tests.test_loads.TestWDRT)
>----------------------------------------------------------------------
>Traceback (most recent call last):
  File "D:\a\MHKiT-Python\MHKiT-Python\mhkit\tests\test_loads.py", line 176, in test_MLERexportTimeSeries
    assert_frame_equal(mler_ts,self.mler_ts)
  File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\site-packages\pandas\_testing\asserters.py", line 1321, in assert_frame_equal
    check_index=False,
  File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\site-packages\pandas\_testing\asserters.py", line 1084, in assert_series_equal
    index_values=np.asarray(left.index),
  File "pandas\_libs\testing.pyx", line 53, in pandas._libs.testing.assert_almost_equal
  File "pandas\_libs\testing.pyx", line 168, in pandas._libs.testing.assert_almost_equal
  File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\site-packages\pandas\_testing\asserters.py", line 665, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: DataFrame.iloc[:, 0] (column name="WaveHeight") are different
>
>DataFrame.iloc[:, 0] (column name="WaveHeight") values are different (100.0 %)
[index]: [-150.0, -149.0, -148.0, -147.0, -146.0, -145.0, -144.0, -143.0, -142.0, -141.0, -140.0, -139.0, -138.0, -137.0, -136.0, -135.0, -134.0, -133.0, -132.0, -131.0, -130.0, -129.0, -128.0, -127.0, -126.0, -125.0, -124.0, -123.0, -122.0, -121.0, -120.0, -119.0, -118.0, -117.0, -116.0, -115.0, -114.0, -113.0, -112.0, -111.0, -110.0, -109.0, -108.0, -107.0, -106.0, -105.0, -104.0, -103.0, -102.0, -101.0, -100.0, -99.0, -98.0, -97.0, -96.0, -95.0, -94.0, -93.0, -92.0, -91.0, -90.0, -89.0, -88.0, -87.0, -86.0, -85.0, -84.0, -83.0, -82.0, -81.0, -80.0, -79.0, -78.0, -77.0, -76.0, -75.0, -74.0, -73.0, -72.0, -71.0, -70.0, -69.0, -68.0, -67.0, -66.0, -65.0, -64.0, -63.0, -62.0, -61.0, -60.0, -59.0, -58.0, -57.0, -56.0, -55.0, -54.0, -53.0, -52.0, -51.0, ...]
[left]:  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...]
[right]: [0.0056111204947132, 0.00676211323244, 0.0065273253096519, 0.0046991101394551, 0.0014999504036406, -0.0024137212076741, -0.0060984982119871, -0.0085865013429153, -0.0091742893672047, -0.0076425632388589, -0.0043371178824015, -7.68944971048072e-05, 0.0040855919102234, 0.0071659349083967, 0.0085133950061872, 0.0079631630401337, 0.0058309545618116, 0.0027662101243936, -0.0004689480960286, -0.0032133230547521, -0.005044737246912, -0.0058167402860824, -0.0056087263493874, -0.0046285747640968, -0.003117685566721, -0.0012959289798035, 0.0006429823127254, 0.0025028285773548, 0.004054649604021, 0.0050393112008922, 0.005221758977708, 0.004481350427664, 0.0028912771678821, 0.0007433958201267, -0.0015027899797669, -0.0033484651238398, -0.0044119024929197, -0.0045371445233994, -0.0038241031654007, -0.0025774692399854, -0.0011886124681957, -1.3421694059798345e-06, 0.00078[222](https://github.com/MHKiT-Software/MHKiT-Python/runs/5321392157?check_suite_focus=true#step:6:222)85566607, 0.0011346804691711, 0.0011788054473029, 0.0011179868166399, 0.001149655861973, 0.0013948847234891, 0.0018607755635891, 0.0024327285891498, 0.0028967741864653, 0.0030008535834661, 0.0025407694746901, 0.0014443471001214, -0.0001771238114199, -0.0020286318167525, -0.0036772175388999, -0.0046535149834697, -0.0046025051869966, -0.0034330195195121, -0.0014201139507787, 0.0007955926230196, 0.0023807524437681, 0.0026310384994883, 0.0013000896288869, -0.0012123707569739, -0.0039324522087599, -0.0056745807314571, -0.0055480438046247, -0.0033621709353109, 0.0002623689437339, 0.0041277864596147, 0.0069614216612002, 0.0079817375122265, 0.0072390205643249, 0.0055694915472529, 0.004198816295025, 0.0041519778162673, 0.0057215652444758, 0.0082527676342736, 0.010341983887752, 0.0103544518122778, 0.0070354995163409, -4.233429893819738e-05, -0.0103026192570156, -0.02[224](https://github.com/MHKiT-Software/MHKiT-Python/runs/5321392157?check_suite_focus=true#step:6:224)91705184736, -0.0337781494775385, -0.04257799279911, -0.0465[230](https://github.com/MHKiT-Software/MHKiT-Python/runs/5321392157?check_suite_focus=true#step:6:230)235582711, -0.044069637016564, -0.0345848441067263, -0.0184807873216028, 0.0028393473321096, 0.0271152985884347, 0.0514365615586159, 0.0725592635690956, 0.0872908209974811, 0.0929586205319039, 0.0879858849964644, 0.07[233](https://github.com/MHKiT-Software/MHKiT-Python/runs/5321392157?check_suite_focus=true#step:6:233)19603541151, ...]

cmichelenstrofer commented:
Error with line 164:

```python
assert_series_equal(mler_norm['WaveSpectrum'], self.mler['Norm_Spec'],check_exact=False,check_less_precise=False,check_names=False)
```

>======================================================================
FAIL: test_MLERwaveAmpNormalize (mhkit.tests.test_loads.TestWDRT)
>----------------------------------------------------------------------
>Traceback (most recent call last):
  File "D:\a\MHKiT-Python\MHKiT-Python\mhkit\tests\test_loads.py", line 164, in test_MLERwaveAmpNormalize
    assert_series_equal(mler_norm['WaveSpectrum'], self.mler['Norm_Spec'],check_exact=False,check_less_precise=False,check_names=False)
  File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\site-packages\pandas\_testing\asserters.py", line 983, in assert_series_equal
    obj=f"{obj}.index",
  File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\site-packages\pandas\_testing\asserters.py", line 355, in assert_index_equal
    _check_types(left, right, obj=obj)
  File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\site-packages\pandas\_testing\asserters.py", line 316, in _check_types
    assert_class_equal(left, right, exact=exact, obj=obj)
  File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\site-packages\pandas\_testing\asserters.py", line 455, in assert_class_equal
    raise_assert_detail(obj, msg, repr_class(left), repr_class(right))
  File "c:\hostedtoolcache\windows\python\3.7.9\x64\lib\site-packages\pandas\_testing\asserters.py", line 665, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: Series.index are different
>
>Series.index classes are not equivalent
[left]:  Float64Index([                  0.0,  0.002004008016032064,
               0.004008016032064128, 0.0060120240480961915,
               0.008016032064128256,   0.01002004008016032,
               0.012024048096192383,  0.0140280561122[244](https://github.com/MHKiT-Software/MHKiT-Python/runs/5321392157?check_suite_focus=true#step:6:244)47,
                0.016032064128[256](https://github.com/MHKiT-Software/MHKiT-Python/runs/5321392157?check_suite_focus=true#step:6:256)51,  0.018036072144288574,
              ...
                 0.9819639278557113,    0.9839679358717434,
                 0.9859719438877754,    0.9879759519038075,
                 0.9899799599198396,    0.9919839679358716,
                 0.9939879759519037,    0.9959919839679358,
                 0.9979959919839678,                   1.0],
             dtype='float64', length=500)
[right]: RangeIndex(start=0, stop=500, step=1)

hivanov-nrel commented:
@rpauly18 @ssolson hi all, Carlos and I were able to finish integrating the WDRT functionalities. Issues were created for future upgrades for whatever we couldnt get to. Sorry for the last minute request, but if you can review the changes we might be able to squeeze this into the next release!

cmichelenstrofer commented:
Closes #22

rpauly18 reviewed (approved):
This looks good. I will note that test_samples_seastate is failing locally for me, but for no apparent reason. Some kind of tolerance setting in my local for assert statements may be off. Great job on this!

ssolson reviewed (commented):
For the sake of time I wanted to quickly share my thoughts around the first example I went over in detail. I will continue my review but thought we could start a discussion around some of this and maybe get some improvements. I want to start by saying that these examples have the fundamentals to be some of the better MHKiT examples, especially from a scientific application point of view. What I find though is that the examples are pretty advanced and lack a lot of explanation that MHKiT examples typically include. Generally speaking, MHKiT caters to an audience who is assumed to be unfamiliar with Python. Many of the functions perform advanced operations without an explanation of what the syntax is doing. 

For instance, the following code block of the short_term_extremes_example assumes the user know what each of the variables are as there is no discussion about what variables are created or why:
```
# short-term period in seconds
t_st = 3 * 60 * 60 

# JONSWAP spectrum for creating short-term time-series
T_min = 1  # s
Tp = 8  # s
Hs = 1.5  # m
f0 = 1/t_st
f_max = 1/T_min
Nf = int(f_max/f0)
f = np.linspace(f0, f_max, Nf)
S = jonswap_spectrum(f, Tp, Hs)

# time in seconds
time = np.linspace(0, t_st, 2*Nf+1)
```
These types of things we typically discuss or point to a reference instead of just listing. I am mostly pointing this out because it is a pattern through the examples.

This is quickly followed by 
`qoi_timeseries = [surface_elevation(S, time).values.squeeze() for _ in range(N)]`
which assumes the user understands we are creating a for loop and not using the underscore.

In the code block following this one the variable `its` is used without explanation.

In the finding, the Global peaks plot section it took me 3 minutes to figure out the difference between the two plots was the x-scale because there is no discussion about what is going on or what I am being shown. 

The "Fit a distribution to the peaks" starts out good but then at the plotting, there is no discussion about the results which Is okay but it would be nice to have some discussion about what the user is looking at and why. 

The example closes by discussing achieving all the results above in one step. First I think this might make more sense to be shown first since it is the most common. Second, I think You should make a plot that demonstrates that we have achieved the same outcome as the plots proceeding the example.

Again I want to repeat that the bones and fundamentals of these examples are really strong but this deviates from a typical MHKiT example. I am okay with moving forward as it can be improved later but I think all of these examples could be improved with a focus on explaining to someone new what is going on.

cmichelenstrofer commented:
@ssolson That makes sense. The short-term extreme example is the most "technical one". I think the full sea-state and contour examples are more approachable (but probably can still be improved in the ways you pointed out).

ssolson commented on mhkit/wave/contours.py:
```diff
@@ -1521,4 +1522,331 @@ def _bivariate_KDE(x1, x2, bw, fit, nb_steps, Ndata_bivariate_KDE, kwargs):
+### Sampling
+def samples_full_seastate(x1, x2, points_per_interval, return_periods,
+                          sea_state_duration, method="PCA", bin_size=250):
+    """ Sample a sea-sate between contours of specified return periods.
```
In MHKiT we tend to use "sea state" with no hyphen. Also check spell.

cmichelenstrofer replied:
Both done w/ @hivanov-nrel latest commit.

ssolson commented on mhkit/wave/contours.py:
```diff
@@ -1521,4 +1522,331 @@ def _bivariate_KDE(x1, x2, bw, fit, nb_steps, Ndata_bivariate_KDE, kwargs):
+    # Sample transformation to principal component space
+    sample_u1 = sample_beta * np.cos(sample_alpha)
+    sample_u2 = sample_beta * np.sin(sample_alpha)
+
+
```
Double spaces are reserved for use between functions

cmichelenstrofer replied:
Done w/ @hivanov-nrel latest commit.

cmichelenstrofer replied:
Done w/ @hivanov-nrel latest commit.

ssolson commented on mhkit/wave/contours.py:
```diff
@@ -1521,4 +1522,331 @@ def _bivariate_KDE(x1, x2, bw, fit, nb_steps, Ndata_bivariate_KDE, kwargs):
+    weight_points: np.array
+        Vector of probabilistic weights for each sampling point
+        to be used in risk calculations.
+    """
+    if method != 'PCA':
```
There are no checks on the types of data passed. These must be added.

cmichelenstrofer replied:
Done w/ @hivanov-nrel latest commit.

ssolson commented on mhkit/wave/contours.py:
```diff
@@ -1521,4 +1522,331 @@ def _bivariate_KDE(x1, x2, bw, fit, nb_steps, Ndata_bivariate_KDE, kwargs):
+    -------
+    hs_samples : nparray
+        points sampled along return contour
+    """
+    #finds minimum and maximum energy period values
```
Add checks on data types and values passed.

cmichelenstrofer replied:
Done w/ @hivanov-nrel latest commit.

ssolson commented on mhkit/wave/contours.py:
```diff
@@ -1521,4 +1522,331 @@ def _bivariate_KDE(x1, x2, bw, fit, nb_steps, Ndata_bivariate_KDE, kwargs):
+        Array of fitted sample radius values.
+    weight_points: np.array
+        Array of weights for each point.
+    """
+    num_samples = (len(beta_lines) - 1) * points_per_interval
```
Internal functions also need to have assert statements to check the incoming data. This allows for changes to function upstream of this to be quickly caught for and errors in data formatting this function may expect.

cmichelenstrofer replied:
Done w/ @hivanov-nrel latest commit.

ssolson commented on mhkit/wave/contours.py:
```diff
@@ -1521,4 +1522,331 @@ def _bivariate_KDE(x1, x2, bw, fit, nb_steps, Ndata_bivariate_KDE, kwargs):
+        # r = rho_zeroline - beta_lines[i + 1]  # TODO
+        r = rho_zeroline - beta_lines[i + 1] + 0.01
+        if any(r < 0):
+            # left = np.amin(np.where(r < -0.01))  # TODO
+            # right = np.amax(np.where(r < -0.01)) # TODO
+            # TODO
```
Why is there a TODO here? Do we still need to do this?

cmichelenstrofer replied:
This was resolved. Need to delete the TODOs and the commented out lines.

cmichelenstrofer replied:
Done w/ @hivanov-nrel latest commit.

ssolson commented on mhkit/wave/contours.py:
```diff
@@ -1521,4 +1522,331 @@ def _bivariate_KDE(x1, x2, bw, fit, nb_steps, Ndata_bivariate_KDE, kwargs):
+        Hs values following rotation from principal component space.
+    original2: np.array
+        T values following rotation from principal component space.
+    """
+    original1 = np.zeros(len(princip_data1))
```
Data checks and assert statements.

cmichelenstrofer replied:
Done w/ @hivanov-nrel latest commit.

ssolson reviewed (commented):
Hey team, having assert statements to check the incoming data is a must-have for all MHKiT functions. Please take a look at my comments here and let me know what we are looking like in terms of time for ensuring the comments here are propagated throughout in terms of style, and constituency in the PR. 

@rpauly18 as a discussion point we may want to consider this PR not ready for the release to ensure there is time for QA on this PR.

cmichelenstrofer commented:
@hivanov-nrel Can we setup a time this or next week to address the review comments (assert statements, example notebooks, etc.). @rpauly18 and @ssolson is there anyone else that can help with these cleanup/style changes? I know @hivanov-nrel and I have very limited time.

ssolson commented:
Realistically it is going to take a new person 2 or 3 times as long to learn the code, make the changes, and then get the changes reviewed by the original authors compared to the original authors making the changes. If you guys are have commitments that prevent you from completing the PR then I can do it but I would not recommend including this PR in a tagged release if we go this way.

hivanov-nrel commented:
I can carve out some time to finish this. Please let us know once the review is complete and Carlos and I can meet to discuss. If we wanted to get this in before the next release, what is the deadline we are aiming for?

ssolson commented:
I would encourage you guys to address general comments around style, spelling, example descriptions, and code cleanup as soon as possible. My understanding is that this PR needs to be wrapped up to tag the release along with Dolfyn. I believe the release should have already gone out.

cmichelenstrofer commented on mhkit/wave/contours.py:
```diff
@@ -1521,4 +1522,331 @@ def _bivariate_KDE(x1, x2, bw, fit, nb_steps, Ndata_bivariate_KDE, kwargs):
+def data_contour():
+    """Create a contour around the data. """
+    raise NotImplementedError()
+
```
Done w/ @hivanov-nrel latest commit.

cmichelenstrofer commented on mhkit/wave/contours.py:
```diff
@@ -1521,4 +1522,331 @@ def _bivariate_KDE(x1, x2, bw, fit, nb_steps, Ndata_bivariate_KDE, kwargs):
+                          sea_state_duration, method="PCA", bin_size=250):
+    """ Sample a sea-sate between contours of specified return periods.
+
+    This function is used for the full seastate approach for the
```
Done w/ @hivanov-nrel latest commit.

hivanov-nrel commented:
@ssolson it looks like Github actions isnt automatically running the tests. Do you know why that would be? It looks like the workflow file wasnt changed and looks like it is set up similar to all the other PRs.

cmichelenstrofer commented:
@hivanov-nrel  I think I know what the issue is. We switched to a PR to the `dev` branch instead of `main`. @rpauly18 and @ssolson if the `dev` branch is now the default the GitHub actions need to be updated to run on PR to `dev` instead  of PR to `main`.

rpauly18 commented:
> @hivanov-nrel I think I know what the issue is. We switched to a PR to the `dev` branch instead of `main`. @rpauly18 and @ssolson if the `dev` branch is now the default the GitHub actions need to be updated to run on PR to `dev` instead of PR to `main`.

Good catch. We should set it to run on both PR to dev and master.

cmichelenstrofer commented:
@hivanov-nrel can you check if the `extreme_response_full_seastate_example.ipynb` and the `extreme_response_contour_example.ipynb` run? I don't have the ability to run these because of the buoy data request

cmichelenstrofer commented:
@ssolson the code is ready for you to continue reviewing. The only thing left is making the examples easier to follow. I will get to that at some point this week. @hivanov-nrel can you take care of the MLER example, I will take care of the other 3.

cmichelenstrofer commented:
@hivanov-nrel  I have cleaned up the contour and full sea state long-term extremes examples. The only examples left to clean are the MLER and the short-term extreme. I will work on the short-term extreme later.

ssolson reviewed (commented):
Carlos, I did a review of `extreme.py` this morning and made a small PR on your branch. I will continue revewing.

cmichelenstrofer commented on mhkit/loads/extreme.py:
```diff
@@ -0,0 +1,718 @@
+    return t_peaks, peaks
+
+
+def npeaks_st(n, t, t_st):
+    """
+    Estimate the number of peaks in a specified period.
```
understood, I tend to use the "nxxx" and "ixxx" but I didn't use it consistently and it doesn't match the rest of MHKiT. I will go with `number_of_short_term_peaks`

hivanov-nrel commented on mhkit/loads/extreme.py:
```diff
@@ -0,0 +1,718 @@
+    Define the simulation parameters that are used in various MLER
+    functionalities.
+
+    See example for how this is useful. If no input is given, then
```
@cmichelenstrofer yes i am referring to the mler example notebook

hivanov-nrel commented:
> @hivanov-nrel I have cleaned up the contour and full sea state long-term extremes examples. The only examples left to clean are the MLER and the short-term extreme. I will work on the short-term extreme later.

@cmichelenstrofer i looked over the MLER example and the coding part is pretty straightforward. I have reached out to Eliot for better clarification regarding the theory, but no response yet. Ill try to follow up, but i think that shouldnt stop us from adding it to the release.

ssolson reviewed (commented):
I looked at 
`examples/extreme_response_contour_example.ipynb`
`examples/short_term_extremes_example.ipynb`

One minor  question. adjustment to STE.

cmichelenstrofer commented:
@ssolson or @hivanov-nrel  can one of you re-run (clear outputs, restart kernel, run all, save, git add/commit/push) the following two examples:

- `extreme_response_full_sea_state_example.ipynb`
- `extreme_response_contour_example.ipynb`

I still haven't been able to figure out how to run those from my computer... I want to make sure they run and that the saved outputs match the new changes.

cmichelenstrofer commented:
@ssolson I have finished cleaning up the `extreme.py` file and the last example `short_term_extremes_example.ipynb`. I will wait for the rest of your review. Thanks!

ssolson commented:
> 
Carlos the full sea state example is failing for me. The other 2 examples all run.
![image](https://user-images.githubusercontent.com/13438942/159954432-163a5698-5681-4d87-aefc-f42192e66c03.png)

cmichelenstrofer commented:
@ssolson the example should work now. Thanks!

cmichelenstrofer commented:
Thanks @ssolson! But no one addressed my comment above. @ssolson @hivanov-nrel can one of you re-run those two test cases and push?

cmichelenstrofer commented:
Also, I only switched to a PR on the `main` branch so that the tests would run, this should have been merged into the `dev` branch. I think it's fine this time.

ssolson commented:
@cmichelenstrofer I re ran the test prior to merging and they worked.

I did not notice this was switched but ultimately I believe this is the correct way. I really dislike the dev branch. There are tagged releases on github and pip installs.

cmichelenstrofer commented:
@ssolson I wanted someone to **push** the examples too, in addition to checking that they run. So that the saved outputs match the new changes.

hivanov-nrel commented:
@ssolson @cmichelenstrofer im running them now and will push the updates directly to the master branch

hivanov-nrel commented:
Agh.....it looks like i dont have permission to push directly to master.... @ssolson do you want me to create another PR for this that you can merge? Or do you just want to run them yourselves and commit to master?

ssolson commented:
@hivanov-nrel could you please submit a PR?

hivanov-nrel commented:
@ssolson, sorry for the delay, i had another appointment. The PR has been submitted, please merge when you can. Thanks.
